lms/djangoapps/courseware/management/commands/fix_student_module_newlines.py

```diff
@@ -0,0 +1,330 @@
+                    newline_course_row.id
+                )
+                return FixResult(record_trimmed=0, data_copied=0, record_deleted=0, capa_merge=0, error=1)
+
+        # If we've reached this point, we just want to keep correct_course_row
+        # and delete the newline_course_row entry.
+        log.info(
+            "Conflict: Choosing data from record with correct course_id (%s) "
+            "and deleting row with newline in course_id (%s) - (%s) %s",
+            correct_course_row.id,
+            newline_course_row.id,
+            newline_course_row.module_type,
+            newline_course_row.module_state_key
+        )
+
+        if dry_run:
+            return FixResult(record_trimmed=0, data_copied=0, record_deleted=1, capa_merge=0, error=0)
+
+        try:
+            with transaction.atomic():
+                self.delete_row(newline_course_row)
+            return FixResult(record_trimmed=0, data_copied=0, record_deleted=1, capa_merge=0, error=0)
+        except DatabaseError:
+            log.exception("Could not delete CSM row %s", newline_course_row.id)
+            return FixResult(record_trimmed=0, data_copied=0, record_deleted=0, capa_merge=0, error=1)
+
+    def delete_row(self, model):
+        """Delete the row and history references to it."""
+        for history_row in BaseStudentModuleHistory.get_history([model]):
+            history_row.delete()
+        model.delete()
+
+    def capa_state_and_grade(self, correct_course_row, newline_course_row):
+        """
+        Return a tuple that is (state, grade, max_grade) to preserve.
+
+        THIS SHOULD ONLY BE CALLED FOR CAPA PROBLEMS.
+
+        Updates to scores could have gotten crossed with updates to state, so
+        that the states and scores do not match up to each other (e.g. you might
+        have correct_course_row with the proper state, but the score
+        corresponding to that state might live in newline_course_row). So we're
+        going to re-derive the scores from the state for capa problems.
+        """
+        ccr_grade, ccr_max_grade = self.grade_for_state(correct_course_row.state)
+        if (ccr_grade, ccr_max_grade) != (correct_course_row.grade, correct_course_row.max_grade):
+            log.info(
+                "-> Correct course row %s has grade %s but should have grade %s",
+                correct_course_row.id,
+                (correct_course_row.grade, correct_course_row.max_grade),
+                (ccr_grade, ccr_max_grade)
+            )
+
+        ncr_grade, ncr_max_grade = self.grade_for_state(newline_course_row.state)
+        if (ncr_grade, ncr_max_grade) != (newline_course_row.grade, newline_course_row.max_grade):
+            log.info(
+                "-> Newline course row %s has grade %s but should have grade %s",
+                newline_course_row.id,
+                (newline_course_row.grade, newline_course_row.max_grade),
+                (ncr_grade, ncr_max_grade)
+            )
```

Comment:
Looking at this more, there seems to be one seldom-used response type for which this code will not give the proper grade. Because capa hates us. I don't think it's going to hit the data we're looking at, but I'll add the check against pre-existing items and extra logging.